File: typescript/src/type-specs.ts
import { stat } from "node:fs/promises";
import path from "node:path";

import type { Diagnostic, Severity } from "./core.js";
import { OkfParserError } from "./core.js";

export const SPEC_SLUG_PLACEHOLDER = "{slug}";

const SEPARATOR_RE = /[\s/]+/gu;
const REMOVED_RE = /[^a-z0-9-]+/gu;
const HYPHEN_RUN_RE = /-{2,}/gu;
const COMBINING_RE = /\p{M}/gu;

/**
 * Derive the filesystem-safe slug of one producer-defined type. Accents and
 * cedillas are removed, whitespace and `/` become hyphens, and every remaining
 * non-alphanumeric character is dropped.
 */
export function typeSlug(conceptType: string): string {
  const stripped = conceptType.normalize("NFKD").replace(COMBINING_RE, "");
  const hyphenated = stripped.trim().toLowerCase().replace(SEPARATOR_RE, "-");
  return hyphenated.replace(REMOVED_RE, "").replace(HYPHEN_RUN_RE, "-").replace(/^-+|-+$/gu, "");
}

/**
 * Render the bundle-relative document path expected for one type, or `null`
 * when the type has no slug at all. A type without a slug is a bundle fact and
 * becomes a diagnostic; only an unusable template is an operator error.
 */
export function specRelativePath(template: string, conceptType: string): string | null {
  if (!template.includes(SPEC_SLUG_PLACEHOLDER)) {
    throw new OkfParserError(
      "OKF_SPEC_TEMPLATE",
      `specification template must contain ${SPEC_SLUG_PLACEHOLDER}: ${JSON.stringify(template)}`,
    );
  }
  const slug = typeSlug(conceptType);
  return slug === "" ? null : template.split(SPEC_SLUG_PLACEHOLDER).join(slug);
}

async function isFile(target: string): Promise<boolean> {
  try {
    return (await stat(target)).isFile();
  } catch {
    return false;
  }
}

/**
 * Report every type in use whose derived specification document is absent. The
 * diagnostic is advisory unless `normative` is set: a bundle mid-adoption
 * legitimately has legacy types without a document.
 */
export async function missingTypeSpecs(
  root: string,
  conceptTypes: Iterable<string>,
  template: string,
  options: { readonly normative?: boolean } = {},
): Promise<readonly Diagnostic[]> {
  const severity: Severity = options.normative === true ? "error" : "warning";
  const unique = [...new Set(conceptTypes)].filter((item) => item !== "").sort();
  const diagnostics: Diagnostic[] = [];
  for (const conceptType of unique) {
    const relative = specRelativePath(template, conceptType);
    if (relative === null) {
      diagnostics.push(
        Object.freeze({
          code: "OKF010",
          severity,
          path: template,
          message: `type ${JSON.stringify(conceptType)} has no slug usable as a document path`,
        }),
      );
      continue;
    }
    if (await isFile(path.resolve(root, relative))) {
      continue;
    }
    diagnostics.push(
      Object.freeze({
        code: "OKF010",
        severity,
        path: relative,
        message: `type ${JSON.stringify(conceptType)} has no specification document`,
      }),
    );
  }
  return Object.freeze(diagnostics);
}

File: src/okf_parser/type_specs.py
"""Derive the specification-document path of a producer-defined type.

The ``OKF010``/``OKF011`` rules that use this path run natively
(``okf-engine/src/specs.rs``); the slug is kept here only for the Python
writers that still name files by type until RFC 0024 phase 4 moves them.
Both implementations are pinned by ``conformance/type-spec-slugs.json``.

The path is *derived* rather than declared. A ``spec:`` frontmatter field would
be a second fact free to disagree with the first, and encoding the path in
``type`` itself would tie identity to layout, so renaming a directory would
invalidate every concept. Deriving keeps ``type`` a stable identity and makes
the document location computable.
"""

from __future__ import annotations

import re
import unicodedata

SPEC_SLUG_PLACEHOLDER = "{slug}"
_SEPARATOR_RE = re.compile(r"[\s/]+")
_REMOVED_RE = re.compile(r"[^a-z0-9-]+")
_HYPHEN_RUN_RE = re.compile(r"-{2,}")


class SpecTemplateError(ValueError):
    """Raised when a specification path template cannot address a type."""


def type_slug(concept_type: str) -> str:
    """Derive the filesystem-safe slug of one producer-defined type.

    Accents and cedillas are removed, whitespace and ``/`` become hyphens, and
    every remaining non-alphanumeric character is dropped.
    """
    decomposed = unicodedata.normalize("NFKD", concept_type)
    stripped = "".join(
        character for character in decomposed if not unicodedata.combining(character)
    )
    hyphenated = _SEPARATOR_RE.sub("-", stripped.strip().lower())
    slug = _HYPHEN_RUN_RE.sub("-", _REMOVED_RE.sub("", hyphenated))
    return slug.strip("-")


def spec_relative_path(template: str, concept_type: str) -> str | None:
    """Render the bundle-relative document path expected for one type.

    Returns ``None`` when the type has no slug at all, which a script or a
    non-Latin script name can produce. That is a bundle fact, so it becomes a
    diagnostic; only an unusable template is an operator error.
    """
    if SPEC_SLUG_PLACEHOLDER not in template:
        message = f"specification template must contain {SPEC_SLUG_PLACEHOLDER}: {template!r}"
        raise SpecTemplateError(message)
    slug = type_slug(concept_type)
    if not slug:
        return None
    return template.replace(SPEC_SLUG_PLACEHOLDER, slug)
